main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala

```diff
@@ -3123,32 +3192,36 @@ object Weeder2 {
       }
     }
 
-    def pickJavaName(tree: Tree): Validation[Name.JavaName, CompilationMessage] = {
+    def pickJavaName(tree: Tree)(implicit sctx: SharedContext): Validation[Name.JavaName, CompilationMessage] = {
       val idents = pickQNameIdents(tree)
       mapN(idents) {
         idents => Name.JavaName(idents, tree.loc)
       }
     }
   }
 
-  private def pickQName(tree: Tree)(implicit sctx: SharedContext): Validation[Name.QName, CompilationMessage] = {
-    mapN(pick(TreeKind.QName, tree))(visitQName)
+  private def pickQName(tree: Tree)(implicit sctx: SharedContext): Option[Name.QName] = {
+    visitQName(pick(TreeKind.QName, tree))
   }
 
-  private def visitQName(tree: Tree)(implicit sctx: SharedContext): Name.QName = {
-    expect(tree, TreeKind.QName)
+  private def visitQName(tree: Tree)(implicit sctx: SharedContext): Option[Name.QName] = {
     val idents = pickAll(TreeKind.Ident, tree).map(tokenToIdent)
-    assert(idents.nonEmpty) // We require at least one element to construct a qname
-    val first = idents.head
-    val ident = idents.last
-    val nnameIdents = idents.dropRight(1)
-    val loc = SourceLocation(isReal = true, first.loc.sp1, ident.loc.sp2)
-    val nname = Name.NName(nnameIdents, loc)
-    Name.QName(nname, ident, loc)
+    if (idents.nonEmpty) {
+      val first = idents.head
+      val ident = idents.last
+      val nnameIdents = idents.dropRight(1)
+      val loc = SourceLocation(isReal = true, first.loc.sp1, ident.loc.sp2)
+      val nname = Name.NName(nnameIdents, loc)
+      Some(Name.QName(nname, ident, loc))
+    } else {
+      // Parser has already reported an error so we just return None
+      // TODO: NeedAtLeastOne? Error?
+      None
+    }
   }
 
   private def pickNameIdent(tree: Tree)(implicit sctx: SharedContext): Validation[Name.Ident, CompilationMessage] = {
-    mapN(pick(TreeKind.Ident, tree))(tokenToIdent)
+    Validation.success(tokenToIdent(pick(TreeKind.Ident, tree)))
```

Comment:
We can now start to remove more Validation in subsequent PRs. Of course, we still have 1 `HardFailure` left.